packages/neuron-wallet/src/block-sync-renderer/sync/connector.ts

```diff
@@ -41,15 +49,147 @@ export interface AppendScript {
   async appendScript(_scripts: AppendScript[]) {
     // do nothing
   }
+
+  constructor({ addresses, nodeUrl, indexerUrl }: { addresses: Address[]; nodeUrl: string; indexerUrl: string }) {
+    this.indexer = new CkbIndexer(nodeUrl, indexerUrl)
+    this.addressesByWalletId = addresses
+      .map(address => AddressMeta.fromObject(address))
+      .reduce((addressesByWalletId, addressMeta) => {
+        if (!addressesByWalletId.has(addressMeta.walletId)) {
+          addressesByWalletId.set(addressMeta.walletId, [])
+        }
+
+        const addressMetas = addressesByWalletId.get(addressMeta.walletId)
+        addressMetas!.push(addressMeta)
+
+        return addressesByWalletId
+      }, new Map<string, AddressMeta[]>())
+
+    this.processNextBlockNumberQueue = queue(async () => this.processTxsInNextBlockNumber(), 1)
+    this.processNextBlockNumberQueue.error((err: any) => {
+      logger.error(`Error in processing next block number queue: ${err}`)
+    })
+
+    this.indexerQueryQueue = queue(async (query: any) => {
+      return await this.collectLiveCellsByScript(query)
+    })
+  }
+
+  public stop(): void {
+    this.pollingIndexer = false
+  }
+
+  protected async processNextBlockNumber() {
+    this.processNextBlockNumberQueue?.push(null)
+    await this.processNextBlockNumberQueue?.drain()
+  }
+
+  protected async getTxHashesWithNextUnprocessedBlockNumber(): Promise<[string | undefined, string[]]> {
+    const txHashCachesByNextBlockNumberAndAddress = await Promise.all(
+      [...this.addressesByWalletId.entries()].map(async ([walletId]) =>
+        IndexerCacheService.nextUnprocessedTxsGroupedByBlockNumber(walletId)
+      )
+    )
+    const groupedTxHashCaches = txHashCachesByNextBlockNumberAndAddress.flat().reduce((grouped, txHashCache) => {
+      if (!grouped.get(txHashCache.blockNumber.toString())) {
+        grouped.set(txHashCache.blockNumber.toString(), [])
+      }
+      grouped.get(txHashCache.blockNumber.toString())!.push(txHashCache)
+
+      return grouped
+    }, new Map<string, Array<IndexerTxHashCache>>())
+
+    const nextUnprocessedBlockNumber = [...groupedTxHashCaches.keys()].sort((a, b) => parseInt(a) - parseInt(b)).shift()
+
+    if (!nextUnprocessedBlockNumber) {
+      return [undefined, []]
+    }
+
+    const txHashCachesInNextUnprocessedBlockNumber = groupedTxHashCaches.get(nextUnprocessedBlockNumber)
+
+    return [nextUnprocessedBlockNumber, txHashCachesInNextUnprocessedBlockNumber!.map(({ txHash }) => txHash)]
+  }
+
+  protected async notifyAndSyncNext(indexerTipNumber: number) {
+    const nextUnprocessedBlockNumber = await IndexerCacheService.nextUnprocessedBlock([
+      ...this.addressesByWalletId.keys(),
+    ])
+    if (nextUnprocessedBlockNumber) {
+      this.blockTipsSubject.next({
+        cacheTipNumber: parseInt(nextUnprocessedBlockNumber),
+        indexerTipNumber,
+      })
+      if (!this.processingBlockNumber) {
```

Comment:
I didn't find where `this.processingBlockNumber` is updated

Reply:
It was updated in `indexer-connector`'s and `light-connector`'s `processTxsInNextBlockNumber` and `notifyCurrentBlockNumberProcessed` functions.

Reply:
Got it